### ISSUE TITLE
`optimize.minimize` jac=True option for constraints?

### ISSUE THREAD

**Issue by rfeinman** [scipy/scipy#13842]
The jac=True option for `optimize.minimize` is very useful when using automatic differentiation utilities because it avoids duplicated computations encountered with a separate jac callable. I am wondering if it would be possible to include this option for nonlinear constraints, for example in the `optimize.NonlinearConstraint` objects passed to trust-constr or the dictionaries passed to SLSQP/COBYLA.

A related but less urgent request: it would great to have similar hess=True and hessp=True options for both the objective and constraint functions (helpful when using autograd infrastructure for the same reasons).

**Comment by tupui** [comment]
Hi, automatic differentiation is now available for all trust-constr optimizers #13775. I guess it solves a part of your request?

**Comment by rfeinman** [comment]
@tupui That's not quite what I mean by automatic differentiation. Finite differences is a numerical _approximation_ to the gradient but ultimately only an approximation. Autograd utilities (like e.g. JAX and PyTorch) give you the real gradient in analytical form.

Here's an example function I might use with `optimize.minimize` (with jac=True) using PyTorch autograd utilities as backend:

```python
import torch

def f_jac(x):
    x = torch.tensor(x, requires_grad=True)
    fval = fun(x)
    jac = torch.autograd.grad(fval, x)[0]

    return fval.detach().numpy(), jac.numpy()
```

Another example function I might use if hessp=True were an option:

```python
def f_jac_hessp(x):
    x = torch.tensor(x, requires_grad=True)
    fval = fun(x)
    jac = torch.autograd.grad(fval, x, create_graph=True)[0]
    def hessp(p):
        p = torch.tensor(p)
        return torch.autograd.grad(jac, x, p, retain_graph=True)[0].numpy()
    
    return fval.detach().numpy(), jac.numpy(), hessp
```

**Comment by tupui** [comment]
Indeed, my bad!
